### DISCUSSION TITLE
Tooltip SyncMethod not working

### DISCUSSION THREAD

**Discussion by LucieSumToZero** [recharts/recharts#3529] [Q&A] [answered]
Hi. 
I have 2 different AreaCharts based on the 2 data set below. The 2 data set arrays have different sizes. I would like to synchronize the tooltips of the 2 AreaCharts based on the category values of the data set. As you can see in the data sets, the values of the category (the x values) are different between the 2 data sets. For instance element 0 in the first data set has category value equal to about  1.828, but there is no corresponding category value in the second data set. 
I thought SyncMethod = 'value' would work with my data sets, but unfortunately it's not working. Is that expected? Would there be an alternative? 
Thanks for you help.

First data set:
![image](https://user-images.githubusercontent.com/67582180/232063599-0d04c4fc-954e-402a-9d2f-c785d9001562.png)

Second data set:
![image](https://user-images.githubusercontent.com/67582180/232060580-b94d3360-7eb7-44bb-8e51-16252c882cc7.png)

**Comment by ckifer** [comment]
Hey @LucieSumToZero 

Would be awesome if you could provide a codesandbox for this (template here https://codesandbox.io/s/synchronized-area-chart-forked-bofeq8). That way we can see how you have things set up and try to suggest some solutions. 

My understanding is that if the categories are not the same then there is no way for the chart to be synced. What is your expected result?

**Comment by LucieSumToZero** [comment]
Hey @ckifer 

Thanks for your answer. 
Please see how I have things set up in this codesandbox:  [https://codesandbox.io/s/synchronized-area-chart-forked-mxkebw?file=/src/App.tsx]
As you can see the categories are not the same but the 2 charts share common range for the x values (the x interval of the second chart is included in the x interval of the first chart). The intention here is to have a tooltip displayed on both charts when we are hovering over this common interval. Do you have a solution? 
Would it be possible to have a continuous tooltip when going across the range of the x axis values instead of a tooltip jumping from one value to the other? Indeed, if the tooltip was continuous, then both charts could be synchronized. 

Thanks

**Reply by Yilun-Sun** [reply, marked as answer]
Hi @LucieSumToZero ! Thanks for the example!
I think you could use `syncMethod` to make it synchronized (do a little hack).

https://user-images.githubusercontent.com/43896664/232755867-c656e9cf-bee2-47a3-be60-93cee2bebbbb.mov

It will calculate the nearby tick in data set 2 using the actived value in data set 1.

### Core Code
```JavaScript
const closest = (arr: number[], n: number) =>
  arr.sort((a, b) => Math.abs(a - n) - Math.abs(b - n))[0];


export default function App() {
  const syncMethodFunction = (_tooltipTicks: [], data: any) => {
    const closestValue = closest(
      data02.map((data) => data.x),
      Number(data.activeLabel)
    );

    return data02.findIndex((data) => data.x === closestValue);
  };

  return (
    <div>
      <h4>A demo of synchronized AreaCharts</h4>
      <AreaChart
        width={500}
        height={200}
        data={data01}
        syncId="anyId"
        margin={{
          top: 10,
          right: 30,
          left: 0,
          bottom: 0
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="x" />
        <YAxis />
        <Tooltip />
        <Area type="monotone" dataKey="y" stroke="#8884d8" fill="#8884d8" />
      </AreaChart>
      <p>Maybe some other content</p>
      <AreaChart
        width={500}
        height={200}
        data={data02}
        syncId="anyId"
        syncMethod={syncMethodFunction}
        margin={{
          top: 10,
          right: 30,
          left: 0,
          bottom: 0
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="x" />
        <YAxis />
        <Tooltip />
        <Area type="monotone" dataKey="y" stroke="#82ca9d" fill="#82ca9d" />
      </AreaChart>
    </div>
  );
}

```

### Screenshot
![image](https://user-images.githubusercontent.com/43896664/232754950-2e1b8eac-8f99-4338-9776-56b309551a23.png)


### CodeSandbox
https://codesandbox.io/s/synchronized-area-chart-forked-rv4ueb?file=/src/App.tsx

**Reply by LucieSumToZero** [reply]
Hi @Yilun-Sun , 

Thanks a lot for this great solution! Thanks a lot for your precious help!

**Comment by LucieSumToZero** [comment]
Hi, when adding the solution to my area charts, it deactivates the tooltip viewbox (but not the cursor). Do you have some advice on how I could make sure to keep the tooltip view bow active?

I have modified the 'syncMethodFunction' as below so that it can use the 'value' and 'category' of the data shown in the post on Apr 14, 2023: 
```
  const closest = (arr: number[], n: number) =>
    arr.sort((a, b) => Math.abs(a - n) - Math.abs(b - n))[0];

  const syncMethodFunction = (_tooltipTicks: [], data: any) => {
    let closestValue = Infinity
      closestValue = closest(
        chartData[0][0].structureData.map((data) => data.category),
        Number(data.activeLabel) // data.activeLabel = x value of the active dot
      );
    return chartData[0][0].structureData.findIndex((data) => data.category === closestValue);
  };
```

Thanks for your answer

**Reply by ckifer** [reply]
What do you mean by viewBox?

**Reply by LucieSumToZero** [reply]
The viewbow is the rectangular area where the hovering values are shown: 
![Screenshot 2024-11-05 162116](https://github.com/user-attachments/assets/1cf21b51-e678-44b9-b64d-31931e304f95)

**Reply by ckifer** [reply]
Hi I'm sorry but I think we'd need a reproduction in order to help. It looks like the codesandbox that @Yilun-Sun made is still working fine
https://codesandbox.io/p/sandbox/synchronized-area-chart-forked-rv4ueb
Please fork that and try to reproduce your issue

**Reply by LucieSumToZero** [reply]
Hi, 
Thanks for your answer. I think I found the reason of the issue: I had the property 'allowDuplicatedCategory' set to false in my AreaChart.
Please see bleow how one of the tooltip viewbox disappear with this setting : 
![image](https://github.com/user-attachments/assets/96bcc8dc-b7d7-4a82-8ce6-989de06da854)

Whereas, I would expect something like in the example done by  @Yilun-Sun
![image](https://github.com/user-attachments/assets/6f1a257c-6355-4dd3-90f0-6b35045822e0)
